### PR TITLE
fix(STONEINTG-689): add PR last update snapshot annotation

### PR DESCRIPTION
**Pull request by jsztuka**
## Maintainers will complete the following section

- [x] Commit messages are descriptive enough ([hints](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/))
- [x] Code coverage from testing does not decrease and new code is covered ([check the PR coverage on codecov](https://app.codecov.io/gh/redhat-appstudio/integration-service/pulls))
- [ ] [Controllers diagrams](https://github.com/redhat-appstudio/integration-service/tree/main/docs) are updated when PR changes controllers code  (if applicable)

### CODE REVIEW

File: controllers/statusreport/statusreport_adapter.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ package statusreport
 import (
 	"context"
 	"fmt"
+
```
Comment (jsztuka) [inline controllers/statusreport/statusreport_adapter.go:22]: IDE loves to add spaces for no reason whatsoever, cant remove it :(

File: gitops/snapshot.go [hunk 2]
```diff
@@ -703,3 +706,21 @@ func RemoveIntegrationTestRerunLabel(adapterClient client.Client, ctx context.Co
 
 	return nil
 }
+
+func GetLatestUpdateTime(snapshot *applicationapiv1alpha1.Snapshot) (time.Time, error) {
+	//specific time format choosen, regular format errors out on parsing
+	dateString := "2005/01/02/15:04:05"
+	latestUpdateTime := snapshot.GetAnnotations()[SnapshotPRLastUpdate]
+	if latestUpdateTime == "" {
+		err := metadata.SetAnnotation(snapshot, SnapshotPRLastUpdate, time.Now().String())
```
Comment (MartinBasti) [inline gitops/snapshot.go:715]: Why Get method should update something?
This is against SRP and it doesn't belong here, setting annotation must be done somewhere else.

File: status/reporters.go [hunk 3]
```diff
@@ -480,6 +480,21 @@ func (r *GitHubReporter) ReportStatusForSnapshot(k8sClient client.Client, ctx co
...
+				return err
+			}
+			timeDiff := latestUpdateTime.Sub(integrationTestStatusDetail.LastUpdateTime)
+			if timeDiff < 0 {
+				err := metadata.SetAnnotation(snapshot, gitops.SnapshotPRLastUpdate, integrationTestStatusDetail.LastUpdateTime.String())
+				if err != nil {
+					logger.Error(err, "failed to set annotation for snapshot",
+						"snapshot.NameSpace", snapshot.Namespace, "snapshot.Name", snapshot.Name)
+					return err
+				}
```
Comment (MartinBasti) [inline status/reporters.go:496]: What's the purpose if this? It has no effect to code execution, it just update timestamp and still executes reporting into GH

**Review by MartinBasti — CHANGES_REQUESTED** [review]

File: status/reporters.go (hunk 3 above)
Comment (jsztuka) [reply status/reporters.go]: Updated

File: status/reporters.go [hunk 4]
```diff
@@ -480,6 +488,14 @@ func (r *GitHubReporter) ReportStatusForSnapshot(k8sClient client.Client, ctx co
 
 		for _, integrationTestStatusDetail := range integrationTestStatusDetails {
 			integrationTestStatusDetail := *integrationTestStatusDetail // G601
+			if latestUpdateTime.Before(integrationTestStatusDetail.LastUpdateTime) {
+				logger.Info("IntegrationTest: %w contains new changes", integrationTestStatusDetail.ScenarioName)
+				if newLatestUpdateTime.Before(integrationTestStatusDetail.LastUpdateTime) {
+					newLatestUpdateTime = integrationTestStatusDetail.LastUpdateTime
+				}
+			} else {
+				logger.Info("IntegrationTest: %w contains no new changes", integrationTestStatusDetail.ScenarioName)
```
Comment (dirgim) [inline status/reporters.go:497]: Should we call a `continue` after this or something similar in order to skip executing creating/updating a checkRun?

Comment (jsztuka) [reply status/reporters.go:497]: Thats what I had there in a first place, it invalid memory address or nil pointer dereference in unit tests. I thought that changes in https://github.com/redhat-appstudio/integration-service/pull/437 should take care of that.

File: status/reporters.go [hunk 5]
```diff
@@ -452,6 +452,14 @@ func (r *GitHubReporter) ReportStatusForSnapshot(k8sClient client.Client, ctx co
 		return fmt.Errorf("sha label not found %q", gitops.PipelineAsCodeSHALabel)
 	}
 	integrationTestStatusDetails := statuses.GetStatuses()
+	latestUpdateTime, _ := gitops.GetLatestUpdateTime(snapshot)
```
Comment (MartinBasti) [inline status/reporters.go:455]: Maybe we should at least to log err in error log when it fail to fetch timestamp

Comment (MartinBasti) [reply status/reporters.go:455]: it's actually there, this line just miss err variable
```suggestion
	latestUpdateTime, err := gitops.GetLatestUpdateTime(snapshot)
```

File: status/reporters.go [hunk 6]
```diff
@@ -452,6 +452,14 @@ func (r *GitHubReporter) ReportStatusForSnapshot(k8sClient client.Client, ctx co
...
+	newLatestUpdateTime := latestUpdateTime
+	if err != nil {
+		logger.Error(err, "failed to get latest update annotation for snapshot",
+			"snapshot.NameSpace", snapshot.Namespace, "snapshot.Name", snapshot.Name)
+		return err
```
Comment (MartinBasti) [inline status/reporters.go:460]: Maybe instead returning error here, we could just use 0 timestamp and let it selfheal

File: status/reporters.go (hunk 4 above)
Comment (MartinBasti) [reply status/reporters.go:497]: if we don't put `continue` here, we don't even need this PR

File: status/reporters.go (hunk 6 above)
Comment (jsztuka) [reply status/reporters.go:459]: Do you agree we keep the error msg? or we want to log it as info?

File: status/reporters.go (hunk 4 above)
Comment (MartinBasti) [reply status/reporters.go:498]: Also we shouldn't log this, it will flood logs with the same msg

File: status/reporters_test.go [hunk 7]
```diff
@@ -557,13 +558,13 @@ var _ = Describe("GitHubReporter", func() {
 
 		It("reports snapshot tests status via CheckRuns", func() {
 			// Create an pending CheckRun
-			hasSnapshot.Annotations["test.appstudio.openshift.io/status"] = "[{\"scenario\":\"scenario1\",\"status\":\"Pending\",\"lastUpdateTime\":\"2023-08-26T17:57:49+02:00\",\"details\":\"pending\"}]"
+			hasSnapshot.Annotations["test.appstudio.openshift.io/status"] = "[{\"scenario\":\"scenario1\",\"status\":\"Pending\",\"lastUpdateTime\":\"2023-08-26T17:57:49Z\",\"details\":\"pending\"}]"
```
Comment (MartinBasti) [inline status/reporters_test.go:561]: Why this change?

File: status/reporters.go (hunk 6 above)
Comment (MartinBasti) [reply status/reporters.go]: error sounds fine

File: status/reporters_test.go (hunk 7 above)
Comment (MartinBasti) [reply status/reporters_test.go:561]: I got it, because it was not updating due timezone mismatch

**Review by MartinBasti — APPROVED** [review]

File: gitops/snapshot.go [hunk 8]
```diff
@@ -703,3 +706,25 @@ func RemoveIntegrationTestRerunLabel(adapterClient client.Client, ctx context.Co
 
 	return nil
 }
+
+func GetLatestUpdateTime(snapshot *applicationapiv1alpha1.Snapshot) (time.Time, error) {
```
Comment (dirgim) [inline gitops/snapshot.go:710]: While we do exercise these functions somewhat in the status/reporters unit tests, I think we'll also need some dedicated unit tests for them to have enough code coverage.

File: gitops/snapshot.go [hunk 9]
```diff
@@ -703,3 +706,25 @@ func RemoveIntegrationTestRerunLabel(adapterClient client.Client, ctx context.Co
...
+	}
+	t := time.Time{}
+	err := t.UnmarshalText([]byte(latestUpdateTime))
+	if err != nil {
+		return t, fmt.Errorf("failed to unmarshal time annotation: %w", err)
+	}
+	return t, nil
+}
+
+func SetLatestUpdateTime(snapshot *applicationapiv1alpha1.Snapshot, t time.Time) error {
```
Comment (dirgim) [inline gitops/snapshot.go:723]: Same here, I think we'll also need some dedicated unit tests for this to have enough code coverage.

**Comment by dirgim** [comment]
nit: the commit message could do with a description of the changes.

**Review by MartinBasti — CHANGES_REQUESTED** [review]
`IsUpdateNeeded` we probably don't need this anymore, as update detection changed.

**Comment by MartinBasti** [comment]
Here is a potential race condition, because creating of lastUpstaded timestamp entry experiences delay between creation and writing the value into k8s, in some cases, we may skip update in case that another test result has been updated around the same time.

Solution would be to keep records of last update per scenario as JSON in the new annotation. Not sure how much will this race condition happen though.

File: gitops/snapshot.go (hunk 8 above)
Comment (jsztuka) [reply gitops/snapshot.go:710]: Added for two cases within existing unit tests, would this coverage suffice?

Comment (dirgim) [reply gitops/snapshot.go:710]: I was thinking more along the lines of adding unit tests to the [gitops/snapshot_test.go](https://github.com/redhat-appstudio/integration-service/blob/main/gitops/snapshot_test.go) to satisfy codecov.

**Comment by MartinBasti** [comment]
> `IsUpdateNeeded` we probably don't need this anymore, as update detection changed.

Still valid

**Comment by jsztuka** [comment]
> > `IsUpdateNeeded` we probably don't need this anymore, as update detection changed.
> 
> Still valid

So even after refactoring there: https://github.com/redhat-appstudio/integration-service/pull/448/ we want to remove it?

**Comment by jsztuka** [comment]
> `IsUpdateNeeded` we probably don't need this anymore, as update detection changed.

removed

**Comment by jsztuka** [comment]
IsUpdateNeeded removed since update detection changed. - Do we want to include that information in commit msg?

File: status/reporters.go [hunk 10]
```diff
@@ -701,10 +704,20 @@ func (r *GitHubReporter) ReportStatusForSnapshot(k8sClient client.Client, ctx co
 	}
 
 	for _, integrationTestStatusDetail := range integrationTestStatusDetails {
+		if latestUpdateTime.Before(integrationTestStatusDetail.LastUpdateTime) {
+			logger.Info("IntegrationTest: %w contains new changes", integrationTestStatusDetail.ScenarioName)
```
Comment (MartinBasti) [inline status/reporters.go:708]:
```suggestion
			logger.Info("Integration Test contains new status updates", "scenario.Name" integrationTestStatusDetail.ScenarioName)
```

Comment (jsztuka) [reply status/reporters.go]: updated

**Review by MartinBasti — APPROVED** [review]

File: status/reporters.go [hunk 11]
```diff
@@ -480,6 +489,15 @@ func (r *GitHubReporter) ReportStatusForSnapshot(k8sClient client.Client, ctx co
 
 		for _, integrationTestStatusDetail := range integrationTestStatusDetails {
 			integrationTestStatusDetail := *integrationTestStatusDetail // G601
+			if latestUpdateTime.Before(integrationTestStatusDetail.LastUpdateTime) {
```
[the remark this answers is not in the document]
Comment (dirgim) [reply status/reporters.go]: Resolving this to unblock merging, #448 has been merged.
